hack/generator/pkg/codegen/pipeline_determine_resource_ownership.go

```diff
@@ -0,0 +1,293 @@
+	resourceName astmodel.TypeName,
+	resourceSpecName astmodel.TypeName,
+	specType *astmodel.ObjectType) (*astmodel.TypeDefinition, error) {
+
+	// We're looking for a magical "Resources" property - if we don't find
+	// one just move on
+	resourcesProp := findResourcesProperty(specType)
+	if resourcesProp == nil {
+		return nil, nil
+	}
+
+	// The resources property should be an array
+	resourcesPropArray, ok := resourcesProp.PropertyType().(*astmodel.ArrayType)
+	if !ok {
+		return nil, errors.Errorf(
+			"Resource %s has spec %s with Resources property whose type is %T not array",
+			resourceName,
+			resourceSpecName,
+			resourcesProp.PropertyType())
+	}
+
+	// We're really interested in the type of this array
+	resourcesPropertyTypeName, ok := resourcesPropArray.Element().(astmodel.TypeName)
+	if !ok {
+		return nil, errors.Errorf(
+			"Resource %s has spec %s with Resources property whose type is array but whose inner type is not TypeName, instead it is %T",
+			resourceName,
+			resourceSpecName,
+			resourcesPropArray.Element())
+	}
+
+	resourcesDef, ok := definitions[resourcesPropertyTypeName]
+	if !ok {
+		return nil, errors.Errorf("couldn't find definition Resources property type %s", resourcesPropertyTypeName)
+	}
+
+	return &resourcesDef, nil
+}
+
+func resolveResourcesTypeNames(
+	resourcesPropertyName astmodel.TypeName,
+	resourcesPropertyType *astmodel.ObjectType) ([]astmodel.TypeName, error) {
+	var results []astmodel.TypeName
+
+	// Each property type is a subresource type
+	for _, prop := range resourcesPropertyType.Properties() {
+		// TODO: Do we need a recursive function here since this can also be a OneOf?
+		optionalType, ok := prop.PropertyType().(*astmodel.OptionalType)
+		if !ok {
+			return nil, errors.Errorf(
+				"OneOf type %s property %s not of type *astmodel.OptionalType",
+				resourcesPropertyName.Name(),
+				prop.PropertyName())
+		}
+
+		propTypeName, ok := optionalType.Element().(astmodel.TypeName)
+		if !ok {
+			return nil, errors.Errorf(
+				"OneOf type %s optional property %s not of type astmodel.TypeName",
+				resourcesPropertyName.Name(),
+				prop.PropertyName())
+		}
+		results = append(results, propTypeName)
+	}
+
+	return results, nil
+}
+
+func extractChildResourceTypeNames(resourcesPropertyTypeDef astmodel.TypeDefinition) ([]astmodel.TypeName, error) {
+	// This type should be ResourceType, or ObjectType if modelling a OneOf/AllOf
+	_, isResource := resourcesPropertyTypeDef.Type().(*astmodel.ResourceType)
+	resourcesPropertyTypeAsObject, isObject := resourcesPropertyTypeDef.Type().(*astmodel.ObjectType)
+	if !isResource && !isObject {
+		return nil, errors.Errorf(
+			"Resources property type %s was not of type *astmodel.ObjectType or *astmodel.ResourceType, instead %T",
+			resourcesPropertyTypeDef.Name(),
+			resourcesPropertyTypeDef.Type())
+	}
+
+	// Determine if this is a OneOf/AllOf
+	// TODO: Checking for the presence of the JSON marshal function is a bit of a hack...
```

Comment:
Agree that this is possibly a good idea -- although I do see some possible stumbling blocks. We could use something like that to identify ARM types vs Kube types too (since at least right now they're in the same package).

It also adds a level of awkwardness too in some ways because it becomes an enum attempting to summarize structure.

Maybe the metadata approach is good, where stages that create new objectTypes can optionally tag them with some metadata, and that metadata will just be used for filtering (so if you just want to act on Kube types, or you just want to act on OneOf types)... for structural stuff you'd still need to check "is this a kube type that has a name"...

I think I like it.

Reply:
Let's push to a follow-on PR though? I don't wanna bloat this one further if I can avoid it.

Reply:
Agreed to postpone this for now - will revisit when we do the `ObjectType` `Flags`/`Metadata`